aptos-move/move-examples/rewards_pool/sources/rewards_pool.move

```diff
@@ -0,0 +1,211 @@
+        let rewards_pool_constructor_ref = &object::create_object(@rewards_pool);
+        let rewards_pool_signer = &object::generate_signer(rewards_pool_constructor_ref);
+        let rewards_pool_addr = signer::address_of(rewards_pool_signer);
+        let epoch_rewards = smart_table::new();
+        vector::for_each(reward_tokens, |reward_token| {
+            let reward_token: Object<Metadata> = reward_token;
+            let store_constructor_ref = &object::create_object(rewards_pool_addr);
+            let store = fungible_asset::create_store(store_constructor_ref, reward_token);
+            smart_table::add(&mut epoch_rewards, reward_token, EpochRewards {
+                amount: smart_table::new(),
+                store,
+                store_extend_ref: object::generate_extend_ref(store_constructor_ref),
+            });
+        });
+        move_to(rewards_pool_signer, RewardsPool {
+            epoch_allocations: smart_table::new(),
+            epoch_rewards,
+            reward_tokens,
+        });
+        object::object_from_constructor_ref(rewards_pool_constructor_ref)
+    }
+
+    #[view]
+    /// Return the amounts of claimable rewards for a given claimer, rewards pool, and epoch.
+    /// The return value is a vector of reward tokens and a vector of amounts.
+    public fun claimable_rewards(
+        claimer: address,
+        rewards_pool: Object<RewardsPool>,
+        epoch: u64,
+    ): (vector<Object<Metadata>>, vector<u64>) acquires RewardsPool {
+        assert!(epoch < epoch::now(), EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH);
+        let rewards_tokens = safe_rewards_pool_data(&rewards_pool).reward_tokens;
+        let non_empty_reward_tokens = vector[];
+        let reward_per_tokens = vector[];
+        vector::for_each(rewards_tokens, |reward_token| {
+            let reward = rewards(claimer, rewards_pool, reward_token, epoch);
+            if (reward > 0) {
+                vector::push_back(&mut non_empty_reward_tokens, reward_token);
+                vector::push_back(&mut reward_per_tokens, reward);
+            };
+        });
+        (non_empty_reward_tokens, reward_per_tokens)
+    }
+
+    /// Allow a claimer to claim the rewards for a past epoch.
+    public fun claim_rewards(
+        claimer: address,
+        rewards_pool: Object<RewardsPool>,
+        reward_token: Object<Metadata>,
+        epoch: u64,
+    ): FungibleAsset acquires RewardsPool {
+        assert!(epoch < epoch::now(), EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH);
+        let reward = rewards(claimer, rewards_pool, reward_token, epoch);
+        assert!(reward > 0, ENO_REWARDS_TO_CLAIM);
+        let reward_store = smart_table::borrow(&safe_rewards_pool_data(&rewards_pool).epoch_rewards, reward_token);
+        let store_signer = &object::generate_signer_for_extending(&reward_store.store_extend_ref);
+        fungible_asset::withdraw(store_signer, reward_store.store, reward)
```

Comment:
It seems you never update the shares/rewards in the EpochRewards.amount for claim_rewards. Also the shares in epoch_allocation should be updated by calling `decrease_allocation` correspondingly. If you update, the pool can be cleaned.

Reply:
Its not correct to decrease allocation after one claim as users can be claiming a single token each time instead of all tokens. With that said though, may be this can just only support claiming all tokens at once 🤔